cpp/arcticdb/processing/clause.cpp

```diff
@@ -25,1148 +27,1279 @@ namespace arcticdb {
+                            std::forward<typename col_type_info::RawType>(val),
+                            std::forward<typename col_type_info::RawType>(group_id)));
+                  } else {
+                    *row_to_group_ptr++ = it->second;
+                  }
                 });
-        } else {
-            util::raise_rte("Expected single column from expression");
-        }
-        row_slices.erase(std::next(it).base());
-    }
-    SegmentInMemory seg;
-    auto index_col = std::make_shared<Column>(make_scalar_type(grouping_data_type), grouping_map.size(), true, false);
-
-    seg.add_column(scalar_field(grouping_data_type, grouping_column_), index_col);
-    seg.descriptor().set_index(IndexDescriptorImpl(0, IndexDescriptorImpl::Type::ROWCOUNT));
 
-    details::visit_type(grouping_data_type, [&grouping_map, &index_col](auto data_type_tag) {
+            num_unique = next_group_id;
+            util::check(num_unique != 0, "Got zero unique values");
+            for (auto agg_data : folly::enumerate(aggregators_data)) {
+              auto input_column_name =
+                  aggregators_.at(agg_data.index).get_input_column_name();
+              auto input_column = row_slice.get(input_column_name);
+              std::optional<ColumnWithStrings> opt_input_column;
+              if (std::holds_alternative<ColumnWithStrings>(input_column)) {
+                auto column_with_strings = std::get<ColumnWithStrings>(input_column);
+                // Empty columns don't contribute to aggregations
+                if (!is_empty_type(column_with_strings.column_->type().data_type())) {
+                  opt_input_column.emplace(std::move(column_with_strings));
+                }
+              }
+              agg_data->aggregate(opt_input_column, row_to_group, num_unique);
+            }
+          });
+    } else {
+      util::raise_rte("Expected single column from expression");
+    }
+    row_slices.erase(std::next(it).base());
+  }
+  SegmentInMemory seg;
+  auto index_col = std::make_shared<Column>(make_scalar_type(grouping_data_type),
+                                            grouping_map.size(), true, false);
+
+  seg.add_column(scalar_field(grouping_data_type, grouping_column_), index_col);
+  seg.descriptor().set_index(
+      IndexDescriptorImpl(0, IndexDescriptorImpl::Type::ROWCOUNT));
+
+  details::visit_type(
+      grouping_data_type, [&grouping_map, &index_col](auto data_type_tag) {
         using col_type_info = ScalarTypeInfo<decltype(data_type_tag)>;
         auto hashes = grouping_map.get<typename col_type_info::RawType>();
         std::vector<std::pair<typename col_type_info::RawType, size_t>> elements;
-        for (const auto &hash : *hashes)
-            elements.push_back(std::make_pair(hash.first, hash.second));
+        for (const auto& hash : *hashes)
+          elements.push_back(std::make_pair(hash.first, hash.second));
```

Comment:
I find single-line for-loops (and if statements) not having curly braces more error prone than with them. But I'm pretty sure @willdealtry feels the opposite 😁

Reply:
I prefer braces as well.